File: apps/plant_shop_haskell_EN_PAUSE/src/Models/Order.hs
{-# LANGUAGE DeriveGeneric     #-}
{-# LANGUAGE OverloadedStrings #-}

module Models.Order where

import           Control.Applicative            ((<|>))
import           Data.Aeson                     (FromJSON (..), ToJSON (..),
                                                 genericParseJSON, object, (.=),
                                                 withObject, (.:?))
import           Data.Aeson.Casing              (aesonDrop, camelCase)
import           Data.Text                      (Text)
import           Data.Time                      (UTCTime)
import           GHC.Generics                   (Generic)
import           Models.OrderItem               (FullOrderItem)
import           Database.PostgreSQL.Simple.FromRow (FromRow (..), field)

-- | Modèle de base pour une commande, tel que stocké en DB.
data Order = Order
  { orderId        :: Int
  , orderUserId    :: Int
  , orderTotal     :: Double
  , orderStatus    :: Text
  , orderCreatedAt :: UTCTime
  } deriving (Show, Generic)

-- | Modèle complet pour une commande, incluant les articles.
-- | C'est cette structure qui est envoyée via l'API.
data FullOrder = FullOrder
  { fullOrderId        :: Int
  , fullOrderUserId    :: Int
  , fullOrderTotal     :: Double
  , fullOrderStatus    :: Text
  , fullOrderCreatedAt :: UTCTime
  , fullOrderItems     :: [FullOrderItem]
  } deriving (Show, Generic)

instance ToJSON FullOrder where
  toJSON fullOrder =
    object
      [ "id"         .= fullOrderId fullOrder
      , "userId"     .= fullOrderUserId fullOrder
      , "total"      .= fullOrderTotal fullOrder
      , "status"     .= fullOrderStatus fullOrder
      , "createdAt"  .= fullOrderCreatedAt fullOrder
      , "orderItems" .= fullOrderItems fullOrder
      ]

-- | DTO pour la création d'une commande.
data CreateOrderPayload = CreateOrderPayload
  { createOrderItems :: [CreateOrderItemPayload]
  } deriving (Show, Generic)

data CreateOrderItemPayload = CreateOrderItemPayload
  { orderItemPlantId  :: Int
  , orderItemQuantity :: Int
  } deriving (Show, Generic)

instance FromJSON CreateOrderItemPayload where
  parseJSON = genericParseJSON (aesonDrop 9 camelCase)

instance FromJSON CreateOrderPayload where
  parseJSON = genericParseJSON (aesonDrop 11 camelCase)

-- | DTO pour la mise à jour du statut d'une commande.
newtype UpdateOrderStatusPayload = UpdateOrderStatusPayload
  { updateOrderStatus :: Text
  } deriving (Show, Generic)

instance FromJSON UpdateOrderStatusPayload where
  parseJSON = withObject "UpdateOrderStatusPayload" $ \obj -> do
    status <- obj .:? "status" <|> obj .:? "orderStatus"
    maybe (fail "Champ 'status' manquant") (pure . UpdateOrderStatusPayload) status

instance FromRow Order where
  fromRow = Order <$> field <*> field <*> field <*> field <*> field

File: apps/plant_shop_java/Structure_monolithique/plant_shop_java/src/model/Order.java
package model;

import java.math.BigDecimal;
import java.sql.Timestamp;

/**
 * Modèle représentant une commande.
 */
public final class Order {
    public int        id;
    public int        userId;
    public BigDecimal total;
    public String     status;
    public Timestamp  createdAt;   // null lors de l’insertion

    /**
     * Constructeur complet pour lecture DB.
     * @param id int Identifiant
     * @param userId int ID utilisateur
     * @param total BigDecimal Total
     * @param status String Statut
     * @param createdAt Timestamp Date création
     */
    public Order(int id, int userId, BigDecimal total,
                 String status, Timestamp createdAt) {
        this.id = id;
        this.userId = userId;
        this.total = total;
        this.status = status;
        this.createdAt = createdAt;
    }

    /**
     * Constructeur pour insertion.
     * @param userId int ID utilisateur
     * @param total BigDecimal Total
     * @param status String Statut
     */
    public Order(int userId, BigDecimal total, String status) {
        this(0, userId, total, status, null);
    }
}
